cql3/restrictions/statement_restrictions.cc

```diff
@@ -1057,6 +1061,359 @@ ::shared_ptr<expression> make_conjunction(::shared_ptr<expression> a, ::shared_p
         }, *a);
 }
 
+namespace {
+
+using cql3::selection::selection;
+
+/// Serialized values for all types of cells.
+struct row_data {
+    const std::vector<bytes>& partition_key;
+    const std::vector<bytes>& clustering_key;
+    const std::vector<bytes_opt>& other_columns;
+};
+
+/// Returns col's value from the fetched data.
+bytes_opt get_value(const column_value& col, const selection& selection, row_data& cells,
+                    const query_options& options) {
+    auto cdef = col.col;
+    if (col.sub) {
+        auto col_type = static_pointer_cast<const collection_type_impl>(cdef->type);
+        if (!col_type->is_map()) {
+            throw exceptions::invalid_request_exception("subscripting non-map column");
+        }
+        const auto deserialized = cdef->type->deserialize(*cells.other_columns[selection.index_of(*cdef)]);
+        const auto& data_map = value_cast<map_type_impl::native_type>(deserialized);
+        const auto key = col.sub->bind_and_get(options);
+        auto&& key_type = col_type->name_comparator();
+        const auto found = with_linearized(*key, [&] (bytes_view key_bv) {
+            return std::find_if(data_map.cbegin(), data_map.cend(), [&] (auto&& element) {
+                return key_type->compare(element.first.serialize_nonnull(), key_bv) == 0;
+            });
+        });
+        return found == data_map.cend() ? bytes_opt() : bytes_opt(found->second.serialize_nonnull());
+    } else {
+        switch (cdef->kind) {
+        case column_kind::partition_key:
+            return cells.partition_key[cdef->id];
+        case column_kind::clustering_key:
+            return cells.clustering_key[cdef->id];
+        case column_kind::static_column:
+        case column_kind::regular_column:
+            return cells.other_columns[selection.index_of(*cdef)];
+        default:
+            throw exceptions::unsupported_operation_exception("Unknown column kind");
+        }
+    }
+}
+
+/// The comparator type for cv.
+data_type comparator(const column_value& cv) {
+    return cv.sub ?
+            static_pointer_cast<const collection_type_impl>(cv.col->type)->value_comparator() : cv.col->type;
+}
```

Comment:
And then, we can pre-compute things like this comparator.

Reply:
Ack.